### PR TITLE
style-check/vera++: add style validator and scripts

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR adds configuration files and a check script for using [Vera++](https://bitbucket.org/verateam/vera/overview) in RIOT. This tool does C/C++ style checks with a given set of rules. The difference to other tools (like Uncrustify) is the fact this is not a code beautifier but a code validator.

New rules can be added in TCL, Python or LUA. The tool defines a "profile" which consists in a set of rules to be applied and some rule-specific parameters. A default `riot` profile was added.

This PR also adds a `exclude` file for excluding source files based on regular expressions (see `--exclusions` command [here](https://bitbucket.org/verateam/vera/wiki/Running))

The `riot` profile enables all rules except:
- F002 File names should be well-formed
- T002 Reserved names should not be used for preprocessor macros
- T011 Curly brackets from the same pair should be either in the same line or in the same column
- T012 Negation operator should not be used in its short form
- T014 Source files should refer the Boost Software License
- T016 Calls to min/max should be protected against accidental macro substitution

And also defines some parameters for certain rules:
- L004: parameter=max-line-length=80
- L005: parameter=max-consecutive-empty-lines=1

I also enabled "T008 Keywords catch, for, if, switch and while should be followed by a single space" but not sure if it's intended. 

All feedback is more than welcome!

Check README.md file for more details.

### Test procedure
Run `dist/tools/vera++/check.sh`

### Issues/PRs references
#8519 (which could be dropped if this one gets merged)

### CODE REVIEW

**Comment by jia200x** [comment]
[This](https://gist.github.com/jia200x/e2f855d14dfac34c887ff35eb68b8530) is how it looks

**Comment by jia200x** [comment]
> From how I interpret the coding conventions, line length  >80 should be a warning, >100 should be an error. Is that possible  to implement with Vera++?

You are right. This is not possible directly, but can be solved adding a `riot-force` profile (that contains critical stuff like line length > 80) and keep the `riot` profile with warnings.

Then, it's required to execute vera++ 2 times (one with warnings and the other one with errors). What do you think?

**Comment by kaspar030** [comment]
So vera complains but cannot not reformat automatically?

**Comment by jia200x** [comment]
>So vera complains but cannot not reformat automatically?

[Yes it can](https://bitbucket.org/verateam/vera/wiki/Transformations). There are some default transformations we might use (trailing whitespaces, etc)

**Comment by kaspar030** [comment]
IMO uncrustify was a good idea. See https://github.com/RIOT-OS/RIOT/pull/8519#issuecomment-413804502.

**Comment by jcarrano** [comment]
@jia200x @miri64 If we are willing to modify the TCL files, the warning on line length >80 and throwing an error on length > 100 should be perfectly possible.

Vera has only one "report" function, and it's up to the user to define a convention for report lines (something like `Warning:`/`Error:` etc). Then we can have two config values and have the script emit either a Warning or and Error.

Also, if I understand correctly, each script is run independently, meaning each file is traversed multiple times?

@jia200x I suggest we include a few example configurations for popular editors, so that people can run Vera/Uncrustify while they are editing the file. It should encourage people to not wait till the last moment to fix the source.

**Comment by jcarrano** [comment]
> However, it's possible to configure these editors to run Uncrustify with a user defined command.

My point is, if vim is so cool as it's followers claim it should be 100% possible to run Vera in "lint mode" (i.e. visually mark lines that are problematic). I don't use vim so I don't know.

**Comment by jia200x** [comment]
> My point is, if vim is so cool as it's followers claim it should be 100% possible to run Vera in "lint mode" (i.e. visually mark lines that are problematic). I don't use vim so I don't know.

Yes, it's possible

**Comment by miri64** [comment]
> My point is, if vim is so cool as it's followers claim it should be 100% possible to run Vera in "lint mode" (i.e. visually mark lines that are problematic). I don't use vim so I don't know.

Should be possible, and I imagine, that popular linter-plugins like `syntastic` already provide a config for vera.

**Comment by jcarrano** [comment]
From the vera README:

>  Vera++ is mainly an engine that parses C++ source files ....

Looking at the API, though, this statement may be a bit exaggerated. AFAICT vera only tokenizes the files. Any parsing is done by the scripts with ad-hoc state machines (like https://github.com/RIOT-OS/RIOT/blob/dde4d8b21f05bc6e043176ce9cbb2a7222a05fbc/dist/tools/vera%2B%2B/rules/M001.tcl#L19)

**Comment by jia200x** [comment]
> Looking at the API, though, this statement may be a bit exaggerated. AFAICT vera only tokenizes the files. Any parsing is done by the scripts with ad-hoc state machines (like 

Yes, I think they refer to "interpreting" the C++ code into tokens. It's similar [to what Uncrustify does](https://github.com/uncrustify/uncrustify/blob/master/src/tokenize.cpp)

**Comment by cladmi** [comment]
What is the testing procedure for this one with https://github.com/RIOT-OS/riotdocker/pull/49 ?

**Comment by jia200x** [comment]
@cladmi I forgot to add it. Just updated it in the description.

For testing, run `dist/tools/vera++/check.sh`

**Comment by jia200x** [comment]
I pushed the v1.2.1 files within the last commit, so it's compatible with the binary package shipped in Ubuntu

**Comment by jia200x** [comment]
@smlng @aabadie @kaspar030 I just fixed an error with the check script and now static tests pass.
Can someone review again?

In theory this would be ready to be merged and would finally add some exhaustive lint checks to our C code.

**Comment by jia200x** [comment]
ping

File: dist/tools/vera++/profiles/riot [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/tclsh
+
+set rules {
+  L004
```
Comment (smlng) [inline dist/tools/vera++/profiles/riot]: this is also in the force profile, to me it seems like either this or the other profile but not both, right?

Comment (smlng) [reply dist/tools/vera++/profiles/riot]: ah sorry, just saw that this goes with the parameter below, so please ignore

**Comment by jia200x** [comment]
> one minor note: in the profile file one cannot infer what a rule does just by its number, e.g. L004, one always needs to lookup the rule file and see code and comments there. Maybe add short comment in the profile which rules are active, e.g.

It makes sense. I will

**Comment by jia200x** [comment]
@smlng done!
I also fixed added some param files (required by the old vera++)

**Comment by jia200x** [comment]
(btw I couldn't add inline comments. They seems to be invalid in the TCL parser :( )

**Comment by jia200x** [comment]
@smlng may I rebase?

**Comment by smlng** [comment]
yes please rebase (and squash if needed)

**Comment by jia200x** [comment]
rebased and squashed!

**Comment by jia200x** [comment]
@smlng all test passed :)

**Comment by jia200x** [comment]
thanks! Can someone merge then? :)

**Comment by smlng** [comment]
done

**Comment by jia200x** [comment]
thanks!

**Comment by benpicco** [comment]
Hm did you check how many errors vera++ reports for the existing code?
It seems like new PRs are now required to clean up existing code if they touch files with style errors.

**Comment by jia200x** [comment]
> Hm did you check how many errors vera++ reports for the existing code?
It seems like new PRs are now required to clean up existing code if they touch files with style errors.

Hmmm if it's being a problem, maybe we can try to set them to warnings and fix those errors in parallel? Then we can re-enable Vera++ errors

**Pull request merged**